packages/periphery/contracts/oracles/UniswapV2Oracle.sol

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity 0.8.19;
+
+import '@openzeppelin/contracts/access/Ownable2Step.sol';
+import '@openzeppelin/contracts/utils/math/Math.sol';
+import '@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol';
+import '@uniswap/v2-periphery/contracts/libraries/UniswapV2OracleLibrary.sol';
+
+import '@marginly/contracts/contracts/interfaces/IPriceOracle.sol';
+
+/// @notice ExampleSlidingWindowOracle with a little modifications https://github.com/Uniswap/v2-periphery/blob/master/contracts/examples/ExampleSlidingWindowOracle.sol
+contract UniswapV2Oracle is IPriceOracle, Ownable2Step {
+  uint256 private constant X96ONE = 79228162514264337593543950336;
+
+  error WrongValue();
+  error WrongGranularity();
+  error WrongWindow();
+  error PairNotFound();
+  error MissingHistoricalObservation();
+  error UnexpectedTimeElapsed();
+  error PairAlreadyExists();
+
+  struct TokenPair {
+    address quoteToken;
+    address baseToken;
+  }
+
+  struct Observation {
+    uint timestamp;
+    uint priceCumulative;
+  }
+
+  struct PairOracleOptions {
+    uint16 secondsAgo;
+    uint16 secondsAgoLiquidation;
+  }
+
+  address public immutable factory;
+  // the desired amount of time over which the moving average should be computed, e.g. 24 hours
+  uint public immutable windowSize;
+  // the number of observations stored for each pair, i.e. how many price observations are stored for the window.
+  // as granularity increases from 1, more frequent updates are needed, but moving averages become more precise.
+  // averages are computed over intervals with sizes in the range:
+  //   [windowSize - (windowSize / granularity) * 2, windowSize]
+  // e.g. if the window size is 24 hours, and the granularity is 24, the oracle will return the average price for
+  //   the period:
+  //   [now - [22 hours, 24 hours], now]
+  uint8 public immutable granularity;
+  // this is redundant with granularity and windowSize, but stored for gas savings & informational purposes.
+  uint public immutable periodSize;
+
+  // pair address converted into number,
+  // when pairKey > 0 means pair.token0 - is base token and pair.token1 - is quote token
+  int256[] public pairKeys;
+
+  // mapping from pairKey to a list of price observations of that pair
+  mapping(int256 => Observation[]) public pairObservations;
+  // mapping from pairKey to options for that pair
+  mapping(int256 => PairOracleOptions) public pairOptions;
```

Comment:
Style guide: https://docs.soliditylang.org/en/v0.8.25/style-guide.html#order-of-layout

Inside each contract, library or interface, use the following order:
1. Type declarations
2. State variables
3. Events
4. Errors
5. Modifiers
6. Functions